src/hypertable_restrict_info.c

```diff
@@ -71,68 +73,128 @@ dimension_restrict_info_create(Dimension *d)
 }
 
 static bool
-dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber strategy, Const *c)
+remove_all_partitions(bool use_or, DimensionRestrictInfoClosed *dri)
 {
-	int64		value = time_value_to_internal(c->constvalue, c->consttype, false);
+	if (!use_or && list_length(dri->partitions) > 1)
+		/* since this is AND the intersection is empty set */
+	{
+		dri->partitions = NIL;
+		return true;
+	}
+	return false;
+}
 
-	switch (strategy)
+static bool
+dimension_restrict_info_open_add(DimensionRestrictInfoOpen *dri, StrategyNumber strategy, List *values, Oid type)
+{
+	ListCell   *item;
+	bool		restriction_added = false;
+
+	foreach(item, values)
 	{
-		case BTLessEqualStrategyNumber:
-		case BTLessStrategyNumber:
-			if (dri->upper_strategy == InvalidStrategy || value < dri->upper_bound)
-			{
-				dri->upper_strategy = strategy;
-				dri->upper_bound = value;
-			}
-			return true;
-		case BTGreaterEqualStrategyNumber:
-		case BTGreaterStrategyNumber:
-			if (dri->lower_strategy == InvalidStrategy || value > dri->lower_bound)
-			{
-				dri->lower_strategy = strategy;
+		Datum		datum = PointerGetDatum(lfirst(item));
+		int64		value = time_value_to_internal(datum, type, true);
+
+		if (value == -1)
```

Comment:
Basically `time_value_to_internal` returns `-1` when in can't convert Datum type to value. IN/ANY/ALL has specific Oid type that we can't convert.

Reply:
But this is a value inside the IN/ANY/ALL, no. Why would it have the OID of the in/any/all

Reply:
> But this is a value inside the IN/ANY/ALL, no. Why would it have the OID of the in/any/all

We pass OID we've got from `Const`. In case of IN/ANY/ALL that's array type. Since we agreed not to optimize time with IN/ANY, I took this shortcut. Alternatively I was thinking to check if OID is array type and return false immediately ( it looked a bit ugly to copy/paste all array OIDs somewhere and check against, especially since we already have a method that check if it's valid time type). I understand that his look a bit weird but I am not strongly on this. Thoughts?

> why don't we just return false from the function at the beginning if(dimValues->useOr && list_length(dimValues->values) > 1)

We can, but it's been taken care of after `if (value == -1)` check. If `list_length(dimValues->values) > 1` means that OID type is array and we will return after trying to convert to value.

Reply:
Sorry what I meant was `if(dimValues->useOr && list_length(dimValues->values) > 1)` before the foreach and then a value == -1 becomes an error. Also the type can go into the DimensionValues and I would make the type be the base value type, not the corresponding array type. I think this is more correct and reads better. It also handle the ANY(one_item) case which is a plus.

Reply:
Also my proposal would handle the ALL case naturally which is also another plus.

Reply:
I like the idea of wrapping type with dim values. To process ANY/ALL we would first need to convert array type into non-array type: that's easy to do but it will add additional `switch` with few hardcoded Oids (not so nice since it's a weird corner case). However  I am not strongly opinionated about it.  Thoughts?